core/src/banking_stage/transaction_scheduler/transaction_packet_container.rs

```diff
@@ -0,0 +1,324 @@
+            id_to_transaction_ttl: HashMap::with_capacity(capacity),
+            id_to_packet: HashMap::with_capacity(capacity),
+        }
+    }
+
+    /// Returns true if the queue is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.priority_queue.is_empty()
+    }
+
+    /// Returns the remaining capacity of the queue
+    pub(crate) fn remaining_queue_capacity(&self) -> usize {
+        self.priority_queue.capacity() - self.priority_queue.len()
+    }
+
+    /// Get a non-consuming iterator over the top `n` transactions in the queue.
+    pub(crate) fn take_top_n(
+        &mut self,
+        n: usize,
+    ) -> impl Iterator<Item = TransactionPriorityId> + '_ {
+        (0..n).map_while(|_| self.priority_queue.pop_max())
+    }
+
+    /// Get packet by id.
+    pub(crate) fn get_packet(&self, id: &TransactionId) -> Option<&DeserializedPacket> {
+        self.id_to_packet.get(id)
+    }
+
+    /// Get mutable packet by id.
+    pub(crate) fn get_mut_packet(&mut self, id: &TransactionId) -> Option<&mut DeserializedPacket> {
+        self.id_to_packet.get_mut(id)
+    }
+
+    /// Get transaction by id.
+    /// Panics if the transaction does not exist.
+    pub(crate) fn get_transaction(&self, id: &TransactionId) -> &SanitizedTransactionTTL {
+        self.id_to_transaction_ttl
+            .get(id)
+            .expect("transaction must exist")
+    }
+
+    /// Take transaction by id.
+    /// Panics if the transaction does not exist.
+    pub(crate) fn take_transaction(&mut self, id: &TransactionId) -> SanitizedTransactionTTL {
+        self.id_to_transaction_ttl
+            .remove(id)
+            .expect("transaction must exist")
+    }
+
+    /// Insert a new transaction into the container's queues and maps.
+    pub(crate) fn insert_new_transaction(
+        &mut self,
+        transaction_id: TransactionId,
+        packet: ImmutableDeserializedPacket,
+        transaction_ttl: SanitizedTransactionTTL,
+    ) {
+        let priority_id = TransactionPriorityId::new(packet.priority(), transaction_id);
+        if self.push_id_into_queue(priority_id) {
+            self.id_to_packet.insert(
+                transaction_id,
+                DeserializedPacket::from_immutable_section(packet),
+            );
+            self.id_to_transaction_ttl
+                .insert(transaction_id, transaction_ttl);
+        }
+    }
+
+    /// Retries a transaction - inserts transaction back into map (but not packet).
+    pub(crate) fn retry_transaction(
+        &mut self,
+        transaction_id: TransactionId,
+        transaction: SanitizedTransaction,
+        max_age_slot: Slot,
+    ) {
+        let priority = self
+            .id_to_packet
+            .get(&transaction_id)
+            .unwrap()
+            .immutable_section()
+            .priority();
+        let priority_id = TransactionPriorityId::new(priority, transaction_id);
+        if self.push_id_into_queue(priority_id) {
```

Comment:
thinking about it, would it be better to get `priority_details` when sanitizing a transaction, and save the `priority` to sanitized_transaction, instead of making it loosely part of `ImmutableDeserializedPacket`?